mt-kahypar/partition/refinement/flows/parallel_construction.cpp

```diff
@@ -498,6 +647,216 @@ FlowProblem ParallelConstruction<GraphAndGainTypes>::constructOptimizedForLargeH
   return flow_problem;
 }
 
+
+template<typename GraphAndGainTypes>
+FlowProblem ParallelConstruction<GraphAndGainTypes>::constructOptimizedForLargeHEsDeterministic(const PartitionedHypergraph& phg,
+                                                                                                const Subhypergraph& sub_hg,
+                                                                                                const PartitionID block_0,
+                                                                                                const PartitionID block_1,
+                                                                                                vec<HypernodeID>& whfc_to_node) {
+  ALWAYS_ASSERT(!PartitionedHypergraph::is_graph);
+  ASSERT(block_0 != kInvalidPartition && block_1 != kInvalidPartition);
+  FlowProblem flow_problem;
+  flow_problem.total_cut = 0;
+  flow_problem.non_removable_cut = 0;
+  _node_to_whfc.clear();
+
+  tbb::parallel_invoke([&]() {
+    _he_to_whfc.clear();
+    _he_to_whfc.setMaxSize(sub_hg.hes.size());
+    tbb::parallel_for(UL(0), sub_hg.hes.size(), [&](const size_t i) {
+      const HyperedgeID he = sub_hg.hes[i];
+      _he_to_whfc[he] = i;
+    });
+  }, [&] {
+    whfc_to_node.resize(sub_hg.numNodes() + 2);
+  }, [&] {
+    _flow_hg.allocateNodes(sub_hg.numNodes() + 2);
+  }, [&] {
+    _identical_nets.reset();
+  });
+
+  if ( _context.refinement.flows.determine_distance_from_cut ) {
+    _cut_hes.clear();
+  }
+
+  initializeWHFCNodesAndCollectPins(phg, sub_hg, block_0, block_1, flow_problem, whfc_to_node);
+
+  const size_t num_buckets = _pins.numBuckets();
+  const HyperedgeID max_hyperedges = sub_hg.hes.size();
+  const HypernodeID max_pins = sub_hg.num_pins + max_hyperedges;
+  _flow_hg.allocateHyperedgesAndPins(max_hyperedges, max_pins);
+  _flow_hg.setNumCSRBuckets(num_buckets);
+  _tmp_hyperedge_buffer.resize(max_hyperedges);
+  vec<TmpDeterministicHyperedge>& tmp_hyperedges = _tmp_hyperedge_buffer.tmp_hyperedges;
+  vec<uint8_t>& valid_hyperedges = _tmp_hyperedge_buffer.valid_hyperedges;
+
+  ds::ConcurrentBucketMap<HashedHyperedgeInformation> hyperedge_hash_map;
+  hyperedge_hash_map.reserve_for_estimated_number_of_insertions(max_hyperedges);
+  _pins.doParallelForAllBuckets([&](const size_t idx) {
+    vec<TmpPin>& pins_of_bucket = _pins.getBucket(idx);
+    if (pins_of_bucket.size() > 0) {
+      std::sort(pins_of_bucket.begin(), pins_of_bucket.end(),
+        [&](const TmpPin& lhs, const TmpPin& rhs) {
+        return lhs.e < rhs.e || (lhs.e == rhs.e && lhs.pin < rhs.pin);
+      });
+
+      HyperedgeID last_he = kInvalidHyperedge;
+      size_t num_hes = 1;
+      size_t num_pins = 0;
+      for ( const TmpPin& pin : pins_of_bucket ) {
+        if ( pin.e != last_he ) {
+          ++num_hes;
```

Comment:
Doesn't this count one too many if `last_he` is not initialized to `pins_of_bucket.front().e` but `num_hes = 1`. I'm saying `num_hes` could be initialized to zero.